src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp

```diff
@@ -15,337 +15,470 @@
-                    case PHASE_SKELETON:
-                        me->SetDisplayId(MODEL_GHOST);
-                        break;
+                    me->SetDisplayId(MODEL_SKELETON);
+                    Talk(SAY_PHASE_2);
                 }
-                bEventInProgress = true;
+                else if (uiPhase == PHASE_SKELETON)
+                {
+                    me->SetDisplayId(MODEL_GHOST);
+                    SetEquipmentSlots(false, EQUIP_UNEQUIP, EQUIP_NO_CHANGE, EQUIP_NO_CHANGE);
+                    Talk(SAY_PHASE_3);
+                }
+                ++uiPhase;
+                LoadEvents();
             }
         }
 
-        void JustDied(Unit* /*killer*/) override
+        void MovementInform(uint32 type, uint32 id) override
         {
-            DoCast(me, SPELL_KILL_CREDIT);
+            if (type != POINT_MOTION_TYPE)
+                return;
 
-            instance->SetData(BOSS_BLACK_KNIGHT, DONE);
+            if (id == 0)
+            {
+                if (Creature* announcer = instance->GetCreature(DATA_ANNOUNCER))
+                    announcer->AI()->SetData(DATA_BLACK_KNIGHT_PRECAST, 0);
+            }
         }
-    };
 
-    CreatureAI* GetAI(Creature* creature) const override
-    {
-        return GetTrialOfTheChampionAI<boss_black_knightAI>(creature);
-    }
-};
+        void JustSummoned(Creature* summon) override
+        {
+            if (summon->GetEntry() == NPC_DESECRATION_STALKER)
+            {
+                summon->CastSpell(summon, SPELL_DESECRATION_DND, true);
+                summon->CastSpell(summon, SPELL_DESECRATION_ARM, true);
+                return;
+            }
+            summons.Summon(summon);
+            summon->AI()->AttackStart(me->GetVictim());
+        }
 
-class npc_risen_ghoul : public CreatureScript
-{
-public:
-    npc_risen_ghoul() : CreatureScript("npc_risen_ghoul") { }
+        void SummonedCreatureDespawn(Creature* summon) override
+        {
+            summons.Despawn(summon);
+        }
 
-    struct npc_risen_ghoulAI : public ScriptedAI
-    {
-        npc_risen_ghoulAI(Creature* creature) : ScriptedAI(creature)
+        void DamageTaken(Unit* /*done_by*/, uint32& damage) override
         {
-            Initialize();
+            if (damage >= me->GetHealth() && uiPhase < PHASE_GHOST)
+            {
+                damage = 0;
+                events.Reset();
+                me->RemoveAllAuras();
+                me->SetHealth(0);
+                me->SetStandState(UNIT_STAND_STATE_DEAD);
+                DoCast(me, SPELL_BLACK_KNIGHT_DIE);
+                for (SummonList::const_iterator itr = summons.begin(); itr != summons.end(); ++itr)
+                {
+                    Creature* ghoul = ObjectAccessor::GetCreature(*me, *itr);
+                    if (ghoul && ghoul->IsAlive() && !ghoul->HasAura(SPELL_GHOUL_EXPLODE))
+                        ENSURE_AI(npc_risen_ghoul::npc_risen_ghoulAI, ghoul->AI())->DoExplode();
```

Comment:
why not just use `DoAction` or `SetData`

Reply:
Actually not a "why not", ENSURE_AI should be avoided at all costs